streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -565,7 +565,7 @@ public void closeCleanAndRecycleState() {
     protected void maybeWriteCheckpoint(final boolean enforceCheckpoint) {
         // commitNeeded indicates we may have processed some records since last commit
         // and hence we need to refresh checkpointable offsets regardless whether we should checkpoint or not
-        if (commitNeeded) {
+        if (commitNeeded || enforceCheckpoint) {
```

Comment:
What if we just removed the check altogether? It's not like updating the changelog offsets is a particularly "heavy" call, we may as well future-proof things even more by just updating the offsets any time.

In fact, why do we even have this weird split brain logic to begin with...it would make more sense to just update the offsets inside the `StreamTask#maybeWriteCheckpoint` and `stateMgr.checkpoint()` methods, no?

Reply:
I thought the same thing but not knowing enough about all the streams internals I thought I'd just go with the most minimal change possible.

Reply:
Fair enough 🙂

Reply:
The reason that I added this check is that `checkpointableOffsets()` can potentially be expensive. I think the fix to have `commitNeeded || enforceCheckpoint` is actually elegant as we did not introduce extra unnecessary overhead much, since it is only true when closing the task.